crates/clvm-utils/src/tree_hash.rs

```diff
@@ -298,314 +358,569 @@ pub fn tree_hash_cached(a: &Allocator, node: NodePtr, cache: &mut TreeCache) ->
                 let rest = hashes.pop().unwrap();
                 let hash = tree_hash_pair(first, rest);
                 hashes.push(hash);
-                cache.insert(original_node, &hash);
+                // non-costed path: store zero cost for this cached entry
+                cache.insert(original_node, &hash, 0);
             }
+            TreeOp::ConsAddCacheCost(_, _) => unreachable!(),
         }
     }
 
     assert!(hashes.len() == 1);
     hashes[0]
 }
 
-pub fn tree_hash_from_bytes(buf: &[u8]) -> Result<TreeHash, EvalErr> {
-    let mut a = Allocator::new();
-    let node = node_from_bytes_backrefs(&mut a, buf)?;
-    let mut cache = TreeCache::default();
-    Ok(tree_hash_cached(&a, node, &mut cache))
+fn subtract_cost(cost_left: &mut Cost, subtract: Cost) -> Result<(), ()> {
+    if subtract > *cost_left {
+        Err(())
+    } else {
+        *cost_left -= subtract;
+        Ok(())
+    }
 }
 
-#[test]
-fn test_tree_hash() {
-    let mut a = Allocator::new();
-    let atom1 = a.new_atom(&[1, 2, 3]).unwrap();
-    let atom2 = a.new_atom(&[4, 5, 6]).unwrap();
-    let root = a.new_pair(atom1, atom2).unwrap();
-
-    // test atom1 hash
-    let atom1_hash = {
-        let mut sha256 = Sha256::new();
-        sha256.update([1_u8]);
-        sha256.update([1, 2, 3]);
-        let atom1_hash = sha256.finalize();
-
-        assert_eq!(tree_hash(&a, atom1).as_ref(), atom1_hash.as_slice());
-        atom1_hash
-    };
-
-    // test atom2 hash
-    let atom2_hash = {
-        let mut sha256 = Sha256::new();
-        sha256.update([1_u8]);
-        sha256.update([4, 5, 6]);
-        let atom2_hash = sha256.finalize();
-
-        assert_eq!(tree_hash(&a, atom2).as_ref(), atom2_hash.as_slice());
-        atom2_hash
-    };
+#[allow(clippy::result_unit_err)]
```

Comment:
here too, I think it's better to return `Option<>` instead of suppressing this warning

Reply:
Done